creator-node/src/services/sync/processSync.js

```diff
@@ -6,6 +6,43 @@ const { saveFileForMultihashToFS } = require('../../fileManager')
 const { getOwnEndpoint, getCreatorNodeEndpoints } = require('../../middlewares')
 const SyncHistoryAggregator = require('../../snapbackSM/syncHistoryAggregator')
 const DBManager = require('../../dbManager')
+const CIDFailureCountManager = require('./CIDFailureCountManager')
+
+/**
+ * Wrapper around `saveFileForMultihash()` with CID failure count tracking
+ * Throws error if CID failure count is below threshold, else returns boolean indicating saved or skipped
+ * @returns {Boolean} skipped
+ */
+async function skippableSaveFileForMultihashToFS ({
+  serviceRegistry, logger, multihash, expectedStoragePath, gatewaysToTry, fileNameForImage = null, SyncMaxCIDFailureCountBeforeSkip
+}) {
+  let skipped = false
+
+  try {
+    await saveFileForMultihashToFS(
+      serviceRegistry, logger, multihash, expectedStoragePath, gatewaysToTry, fileNameForImage
+    )
+
+    // Reset CID failure count on success
+    CIDFailureCountManager.resetCIDFailureCount(multihash)
+
+    // If saveFileForMultihashToFS throws, it is because of failed content retrieval or verification
+  } catch (e) {
+    // Increment CID failure count
+    CIDFailureCountManager.incrementCIDFailureCount(multihash)
+
+    // If CID failure count is below threshold, throw an error (base case)
+    if (CIDFailureCountManager.getCIDFailureCount(multihash) <= SyncMaxCIDFailureCountBeforeSkip) {
+      throw new Error(e)
```

Comment:
My concern here is tracking the CID individually vs the syncs for the user as a whole. In this code each cid will get retried x times before skipping so effectively you'll need to do cids skipped*num retries to actually complete a sync right?

Reply:
Like we need this code to know if this CID is skipped but I think we should track skip on a per user level

Reply:
^ not sure i follow this comment. if the retries is 2, wouldn't you only need to sync twice to skip?

Reply:
retry twice per cid, so if there's 50 missing cid's that's 100 syncs. correct me if i'm wrong @SidSethi

Reply:
yeah what happens currently is -- failureThreshold = 2, for 50 CIDs -> if any of CIDs has not reached failureThreshold, the `saveFileForMultihash()` call will throw, meaning the outside `await Promise.all()` will fail and reject sync. sync will only succeed if every single skipped CID has hit failure threshold and does not throw an error

Reply:
right but if everything fails in a sync and gets a skip increment, won't that only be two syncs? why would it be 100?

Reply:
since this throws an error, the sync would go into the catch block and that would be the end of it i believe